### ISSUE TITLE
Relationship between ARIA attributes, Accessible Properties and Computed Properties.

### ISSUE THREAD

**Issue by alice** [WICG/aom#60]
Given:
```html
<span id="foo" role="button">
```

What should the interaction be between the AOM object and the ARIA DOM attributes? i.e.: 

```js
var foo = document.getElementById('foo');

/* 1. Should ARIA attributes be reflected to AOM object properties? */
console.log(foo.role);  // What does this print? 

/* 2. Should AOM object properties reflect back out to ARIA attributes */
foo.role = "checkbox";  // Does this affect the DOM or not?
```

And, what should the computed role of `foo` be in this instance? 
- Should ARIA take precedence, i.e. `foo`'s role should stay `button`?
- Or, should AOM take precedence, i.e. `foo`'s role should be `checkbox` after being set on `foo`'s `accessibleNode`?

## Should ARIA DOM attributes be reflected to AOM object properties?
| **Pro** | **Con** 
|-|-| 
| Can read ARIA attributes without using `getAttribute()`. | May set expectation that this is getting the computed value (it won't be). |
| Can be used to [validate/feature detect](https://github.com/WICG/aom/issues/60#issuecomment-287687027) ARIA property values  | |

## Should AOM object properties be reflected to ARIA DOM attributes?
| **Pro** | **Con** |
|-|-|
| More "backwards compatible", doesn't break expectations that ARIA tells the whole story. | Sprouting ARIA attributes is annoying. ARIA/AOM should be implementation details. |
| If things are styling based on ARIA, things still Just Work. | Means that this "shouldn't" be used in Custom Element constructors. |
| Improves ergonomics of using ARIA attributes. | [Performance implications](https://github.com/WICG/aom/issues/60#issuecomment-286956084) with changing attributes triggering style recalc/repaint. |
| | [Not always feasible](https://github.com/WICG/aom/issues/60#issuecomment-286956084) e.g. for relationships.|

## Should AOM or ARIA take precedence?

| **AOM** | **ARIA** |
|-|-|
| Consistent with `input` where [attributes represent only "initial" state](https://github.com/WICG/aom/issues/60#issuecomment-287201660) and properties represent ground truth | Consistent with [philosophy of `style` attribute](https://github.com/WICG/aom/issues/60#issuecomment-286978061), where page author gets the final say via ARIA |
|Avoid forcing developers to [unset ARIA properties](https://github.com/WICG/aom/issues/60#issuecomment-286966859) to be sure AOM changes take effect | Avoid confusion with [ARIA "mysteriously" not working](https://github.com/WICG/aom/issues/60#issuecomment-287183115), although this could be solved by [best practices](https://github.com/WICG/aom/issues/60#issuecomment-287201660)|
| Avoid [diminishing returns](https://github.com/WICG/aom/issues/60#issuecomment-287029155) when compared to a simple reflection of ARIA | |

**Comment by minorninth** [comment]
My opinion is that the safest solution is "no reflection".

One small thing we could do is that setting an AOM property could remove its associated ARIA attribute, to minimize developer confusion. I see no harm in that but I also think it'd be fine if we did nothing.

The next safest thing we could do is reflect ARIA DOM attributes to AOM object properties. That also seems safe, in that you could read ARIA attributes and validate them. Again, I think it'd be fine if we didn't, but it seems perfectly consistent and safe if we did.

Reflecting object properties back onto ARIA attributes seems like a bad idea to me. One additional Con is that it wouldn't work for all properties, like IDREFs, which could be confusing. What would you set the ARIA attribute to then?

**Comment by cookiecrook** [comment]
> My opinion is that the safest solution is "no reflection".

+1. Among other problems, reflections might cause repaints if selectors refer to them. If that behavior is desirable to the author, they can use standard DOM+ARIA interfaces, but I would not want there to be visible or rendering performance side effects of using AOM.

> One small thing we could do is that setting an AOM property could remove its associated ARIA attribute, to minimize developer confusion.

I like this a little better, but it could still cause style recalculation or other side effects. Perhaps the spec could have an RFC-2119 that user agent dev tools MAY or SHOULD throw a warning to the console when a content attribute was change after the corresponding AOM property was defined. 

> The next safest thing we could do is reflect ARIA DOM attributes to AOM object properties. That also seems safe, in that you could read ARIA attributes and validate them. Again, I think it'd be fine if we didn't, but it seems perfectly consistent and safe if we did.

I remembered us agreeing that this was scoped to Stage 4.

> Reflecting object properties back onto ARIA attributes seems like a bad idea to me. One additional Con is that it wouldn't work for all properties, like IDREFs, which could be confusing. What would you set the ARIA attribute to then?

+1

**Comment by alice** [comment]
My position is that ARIA and AOM should be considered completely separate concerns which affect the same end result. i.e.:

```js
el.setAttribute('role', 'grid');
console.log(el.getAttribute('role')); // grid
el.accessibleNode.role;  // null

el.accessibleNode.role = 'treegrid';
console.log(el.accessibleNode.role);  // treegrid
console.log(el.getAttribute('role'));  // grid
```

The issue of confusion with out of date ARIA attributes I think is best addressed by having ARIA attributes take precedence when determining the computed property. This also means that people can continue to use ARIA without being aware of AOM and things will work as expected even if developers use AOM under the hood somewhere, such as with a custom element.

**Comment by cookiecrook** [comment]
> This also means that people can continue to use ARIA without being aware of AOM and things will work as expected even if developers use AOM under the hood somewhere, such as with a custom element.

That feels slippery to me. Potential ramp-up scenario:

- Junior Dev writes a content attribute in markup. 
- shared framework overrides it in AOM (accidentally or otherwise)
- Junior Dev confused why it's not working. Overrides it via `setAttribute` (think `!important`)
- shared framework overrides it again at a later time?
- Junior Dev adds setAttribute throughout the app to keep it "working" (code now unmanageable)

One of the goals of AOM was to reduce the amount of DOM operations in large apps, but if I'm understanding you correctly, this will force engineers to set the AOM property **and** unset (via `removeAttribute`) any related ARIA attributes to ensure everything works.

**Comment by cookiecrook** [comment]
Actually maybe I misunderstood you comment. Clarifying:

```js
el.accessibleNode.role = 'treegrid';
el.setAttribute('role', 'grid');
console.log(el.accessibleNode.role);  // treegrid
console.log(el.getAttribute('role'));  // grid
```
I concur with your example above. I thought you were saying the later `setAttribute` would affect the computedRole in the UA. My opinion is that the computed role should be `treegrid`because the AOM property value overrides the ARIA content attribute value.

**Comment by alice** [comment]
Yes, I was saying that the computed role should be `grid`, because ARIA should take precedence.

**Comment by alice** [comment]
See my clarification PR #64

**Comment by alice** [comment]
> One of the goals of AOM was to reduce the amount of DOM operations in large apps, but if I'm understanding you correctly, this will force engineers to set the AOM property and unset (via removeAttribute) any related ARIA attributes to ensure everything works.

If they want to override ARIA, they can just change the ARIA.

ARIA taking precedence allows e.g. custom element authors to set default semantics, and web page authors to override them using ARIA as they would with a native element, without needing to worry about AOM.

**Comment by esprehn** [comment]
Having aria take precedence matches what we do for style ex.

```html
<my-component style="color: red">
  #shadow-root
     <style>:host { color: blue; }</style>
```

This element is red, because we assume the consumer of the element adding the attribute knows better (they're in the future) as they're attempting to override what the author of the component did (in the past).

That makes sense to me here, where we're saying the consumer of the component can override the internal computed state with an attribute, while component authors should avoid using attributes and should use the AOM instead.

Perhaps I'm not following some part of the AOM model though and it makes sense to diverge?

**Comment by cookiecrook** [comment]
I see these as being different from the style cascade. Ultimately having a directly settable API (and eventually a queryable API for computed values) will be more predictable and verifiable. If content attributes win, the author would need to double check DOM heuristics (or always unset them) to verify the AOM setters are going to be used. 

Another goal of Phase 4 was to be able to check computed values (once privacy and/or perf concerns could be addressed). In the meantime, we discussed it would be sufficient to have accessibleNode respond like a simple dictionary in that you could always read back what you wrote into it. The author would be assured that the accessibleNode values would be the "truth" if and when accessibility was needed. If that dictionary is no longer the source of truth, is there much benefit to the nearly 1:1 ARIA property matching? Why not just add some reflecting DOM properties like the old Microsoft Element interfaces: `el.ariaLabel`, etc?

How would the "ARIA takes precedence" change work with a IDREF vs NodeRef conflict? Which one of these wins? Is it still ARIA? I'd lean towards no.

```js
myDiv.setAttribute('aria-labelledby', 'foo'); // IDREF to <el id=foo>
myDiv.accessibleNode.labelElement = myNodeRef;
```

I'm open to the change (I think), but there may be more implications we haven't considered. It's possible the others I've listed here may be inconsequential.

**Comment by alice** [comment]
I think we should walk through the use cases we have in mind where these situations might arise.

I think Elliott and I are both primarily thinking about a Custom Element type case, where a Custom Element author specifies the default semantics for their element via AOM. In this case, we really want web page authors to be able to override those default semantics using ARIA, for consistency with native elements.

Imagine if you did something like 

```html
<custom-toggle role="switch"></custom-toggle>
```

and `custom-toggle`'s default role of `checkbox` was still applied - it would be a weird and frustrating experience.

It seems like the cases @cookiecrook has in mind are when things are less clear cut, but I don't have a good handle on why the desired semantics might be in conflict in the "Junior dev/shared framework" scenario.

**Comment by CountOrlok** [comment]
I agree in that ARIA should have the last word. If AOM would take precedence, there'd be a need for another mechanism to implement default semantics for Custom Elements, adding even more complexity. That's not desirable.

**Comment by robdodson** [comment]
I think attributes are useful for initial configuration, but beyond that you can't trust them and properties become the source of truth.

An example would be `<input type="checkbox" checked>`. If you select this element and say `.checked=false`, it will render in an unchecked state, though the (now stale) `[checked]` attribute remains in the DOM.

This is one (among many) reasons why frameworks like React and Angular tend to prefer binding to properties instead of attributes.

Reflecting the ARIA attribute to the property sounds good to me. In fact I would _really_ want this. But I wouldn't want or expect a change to the property to reflect back to the attribute, as that's not consistent with how other elements seem to work.

In the Custom Element case, the author _should_ be checking for the presence of an attribute before they programmatically set the role:

```
if (!this.getAttribute('role')) {
  this. accessibleNode.role = 'checkbox';
}
```

Again, this follows on from the idea of attributes for initial configuration.

Similar to how `checked` works today, if someone were to call `setAttribute('role', 'foo')` I would expect the `accessibleNode.role` property to now equal 'foo'.

**Comment by alice** [comment]
I think reflecting ARIA attributes to AOM properties sets the dangerous expectation that AOM properties reflect the *computed* properties. Getting the computed properties carries a potentially heavy performance penalty, so we have agreed to keep that totally separate.

Most reflected attributes reflect in both directions AFAIK - `input` is well known to be weird and confusing, so I'm not sure we want to follow that model.

However, I agree that it's also weird to have two **different** things with the **same** vocabulary and affecting the **same** outcome.

CSS stylesheets vs `.style` vs `getComputedStyle()` is about the only reasonable analogy I can come up with here - however, as @cookiecrook points out, `AccessibleNode`s don't live in a separate domain and match across a whole tree, but are directly attached to a node.

I tried to sum up pro/con arguments again in the top post - feel free to edit there!

**Comment by alice** [comment]
Note: the [explainer](https://github.com/WICG/aom/blob/master/explainer.md) currently reflects the "no reflection, ARIA wins" opinion, but we can update that if we come to a different conclusion here.

**Comment by alice** [comment]
Ok, after talking with @robdodson a little more about this, here's something I think might work better:

The mental model is that AOM/`accessibleNode` reliably reflects the truth about the _author-set_ accessibility properties.

So:
- If ARIA attributes have been used, are valid, and have not been overridden, they are reflected on the `accessibleNode` object:
```html
<div id="composeBtn" role="button">Valid role</div>
<div id="banana" role="banana">Invalid role</div>
<div id="toggle" aria-pressed="true">Valid aria-pressed value in invalid context</div>
```

```js
console.log(composeBtn.accessibleNode.role);  // "button"
console.log(banana.accessibleNode.role);  // null
console.log(toggle.accessibleNode.pressed);  // null
```

- If the author has set a valid property on the `accessibleNode` object, it now becomes the source of truth as far as author-set properties go (AOM wins):
```js
composeBtn.accessibleNode.role = "link";
console.log(composeBtn.accessibleNode.role);  // "link"
```

- This does not reflect _back_ to the ARIA attributes in question, for all the reasons discussed earlier:
```js
console.log(composeBtn.getAttribute("role"));  // "button"
```

So we do lose the "ARIA is a source of truth" guarantee, but at least you actually never need to use `getAttribute()` to check what ARIA has been set on a node, since you can get it via `accessibleNode`, _and_ at the same time check whether it has been overridden.

And, as Rob pointed out, we don't really need to _prevent_ web component authors from overriding web page authors' ARIA, even if it's something they _shouldn't_ do. So a pattern for setting component semantics might look something like:
```js
if (!this.accessibleNode.role)
  this.accessibleNode.role = "checkbox";
```

Finally , the "author-provided properties" mental model explains why you don't get the computed attribute values. Furthermore, if you get a non-null value back from `accessibleNode` you can (I think) be fairly sure that will become the computed value, since it would be valid and thus override any implicit value.

**Comment by minorninth** [comment]
One pro of this idea (reflect ARIA into AOM) is that AOM can be used to *validate* ARIA. It doesn't allow you to check the final computed properties, but it would allow you, for example, to determine if you tried to set an ARIA attribute to an invalid value, or one that the current browser doesn't yet support (feature detection).

I'm still okay either way, interested to hear what @cookiecrook thinks of this last argument before we make a consensus decision.

**Comment by alice** [comment]
> One pro of this idea (reflect ARIA into AOM) is that AOM can be used to validate ARIA.

Added to the Pro/Con table at the top.

**Comment by jakearchibald** [comment]
Given that this stuff can be set via attributes, properties, but has a computed property, aligning with `.style` etc makes to me. That means:

* Attribute getters/setters work without any kind of validation.
* Property setters *and* getters are validated/corrected (as in, incorrect attributes won't be reflected).
* Both are different to computed values, meaning getting doesn't trigger a recalc/layout.

Here's a test case http://jsbin.com/kokemuv/edit?css,js,console.

Some thoughts:

## Setting style properties *does* update the style attribute

```js
div.style.color = 'green';
console.log(div.getAttribute('style'));
// logs "color: green;"
```

The alternative model is like the `.value` attribute on form fields, where the attribute is the source of truth until the DOM property is set or the user changes the value, in which case the DOM property becomes the source of truth. Demo: http://jsbin.com/koxizag/edit?js,console.

Is it worth being different here?

## Unrelated values are validated on set

Given:

```html
<div style="background: green; display: foo"></div>
```

`div.style.display` will return `""` as the value is validated, but `div.getAttribute('style')` will return `background: green; display: foo`. If you execute `div.style.background = 'blue'`, the whole attribute will be validated, so `div.getAttribute('style')` will now return just `background: blue;`.

This seems kinda weird, but if you're going for the same model it might be worth copying.

## When is a value validated?

```html
<div aria-labelledby="foo">…</div>
<div id="foo">…</div>
```

The above is valid ARIA, even though the `aria-labelledby` attribute is created by the parser before `div#foo` exists. Given that the properties return arrays of DOM elements, they work kinda differently:

```html
<div aria-labelledby="foo">…</div>
<script>
  const previousDiv = document.querySelector('[aria-labelledby]');
  previousDiv.setAttribute('aria-labelledby', previousDiv.getAttribute('aria-labelledby'));
</script>
<div id="foo">…</div>
```

I assume this won't change anything - the attribute can be set to itself without an issue. However:

```html
<div aria-labelledby="foo">…</div>
<script>
  const previousDiv = document.querySelector('[aria-labelledby]');
  previousDiv.accessibleNode.labelledBy = previousDiv.accessibleNode.labelledBy;
</script>
<div id="foo">…</div>
```

Will the above set `labelledBy` to an empty list? If validation occurs on 'getting' (as it does with `.style`) then `previousDiv.accessibleNode.labelledBy` will return an array of zero elements.

**Comment by alice** [comment]
> Setting style properties does update the style attribute ...
> Is it worth being different here?

I think the main issue is with relationships/IDREFs. You can set a relationship property via AOM that _cannot_ be represented in the DOM, either for an element with no ID or for an element whose ID is in a different scope (e.g. Shadow DOM). Given we can't reflect _everything_, it seems better not to reflect anything (like `value`).

> Will the above set labelledBy to an empty list?

Not sure whether it'll be an empty list or `null`, but yeah it'll override the previous value with an empty value.

Honestly I think this is ok, though. If you're trying to read things before they exist, you're always going to have a bad time, and I think people will figure this out. It'd be the same if you had a stylesheet at the end of the page and tried to set an element's style to its computed style in the middle of it.

**Comment by jakearchibald** [comment]
> Given we can't reflect everything, it seems better not to reflect anything (like value).

Yeah that's fair. +1 to following `.value`'s weirdness rather than `.style`'s weirdness in this case. It'll need something similar to the [dirty value flag](https://html.spec.whatwg.org/multipage/forms.html#concept-fe-dirty), which is set to change the source of truth.

> Not sure whether it'll be an empty list or `null`

I guess it depends if there's a difference between `<div>` and `<div aria-labelledby="">`.

**Comment by alice** [comment]
> It'll need something similar to the dirty value flag, which is set to change the source of truth.

Good point. @minorninth if we go in this direction we should probably copy that language for the spec.

**Comment by cookiecrook** [comment]
@alice 
> Given we can't reflect everything, it seems better not to reflect anything (like value).

I agree, but I didn't follow you as far as "ARIA always wins over AOM." Perhaps Jake's point will resolve this discrepancy:

@jakearchibald 
> It'll need something similar to the dirty value flag

If I understand this correctly, this would mean we'd follow a "most recent setter wins" pattern, with no reflection back to the DOM. The AOM property is the source of truth for both validation and current value (set in either way), but changes to it are not reflected back into the content attributes.

```js
el.setAttribute("role", "button");
el.accessibleNode.role; // "button" (DOM setter affects AOM)
el.accessibleNode.role = "link"; // role truth is now link, but not reflected back to @role
el.getAttribute("role"); // "button" but now _dirty_ b/c AOM role is "link"
el.setAttribute("role", "note"); // DOM setter now squashes both
el.accessibleNode.role; // "note" (AOM reflects any valid value set after the last time it was set)

el.setAttribute("role", "button link"); // space-separated tokens indicate fallback roles
el.accessibleNode.role; // "button" (first recognized, supported role token)

el.setAttribute("role", "switch checkbox");
el.accessibleNode.role; // "switch" where ARIA 1.1 switch role is supported; "checkbox" elsewhere.
```

If that's the gist of what you want, I'm on board, but what should happen here?

```js
el.setAttribute("role", "foo");
el.accessibleNode.role; // "" (invalid value)
el.setAttribute("role", "button");
el.accessibleNode.role; // "button" (valid value from content attribute)

el.accessibleNode.role = "foo"; // invalid
el.accessibleNode.role; // is this "" (invalid AOM setter) or "button" (fallback value from content attr)?
```

**Comment by alice** [comment]
@cookiecrook Please take a look at [this comment](https://github.com/WICG/aom/issues/60#issuecomment-287607857).

**Comment by jakearchibald** [comment]
@cookiecrook 

> ```js
> el.setAttribute("role", "note"); // DOM setter now squashes both
> ```

`.value` works differently. With `.value`, changes to the attribute are ignored once the value has been altered via the DOM property or by the user, but they're honoured until that point. See http://jsbin.com/koxizag/edit?js,console

**Comment by alice** [comment]
> If I understand this correctly, this would mean we'd follow a "most recent setter wins" pattern, with no reflection back to the DOM. 

Hm, this isn't what I had in mind. I was thinking more that once you set something on AOM, ARIA ceases to be effective. However, I can see that setting ARIA should potentially update the AccessibleNode as well.

> The AOM property is the source of truth for both validation and current value (set in either way), but changes to it are not reflected back into the content attributes.

Yes, this is true.

> If that's the gist of what you want, I'm on board, but what should happen here?

```js
el.setAttribute("role", "foo");
el.accessibleNode.role; // "" (invalid value)
el.setAttribute("role", "button");
el.accessibleNode.role; // "button" (valid value from content attribute)

el.accessibleNode.role = "foo"; // invalid
el.accessibleNode.role; // is this "" (invalid AOM setter) or "button" (fallback value from content attr)?
```

This is still an open question.

I would expect that the invalid AOM property takes precedence, and we fall back on the default semantics of the element (and `el.accessibleNode.role` would return `null` rather than "").

**Comment by alice** [comment]
Collided with @jakearchibald. It might be simpler to follow the `.value` pattern and have ARIA changes do nothing once the `AccessibleNode` has been written to.

**Comment by jakearchibald** [comment]
@cookiecrook 

> ```js
> el.accessibleNode.role = "foo"; // invalid
> el.accessibleNode.role; // is this "" (invalid AOM setter) or "button" (fallback value from content attr)?
> ```

In this case, `.style` falls back to the previous value. Demo: http://jsbin.com/zumivok/edit?js,console

**Comment by cookiecrook** [comment]
@jakearchibald wrote:
> …changes to the attribute are ignored once the value has been altered via the DOM property or by the user

My understanding is that "DOM property" refers to `.value` in this case, and "content attribute" refers to getAttribute("value"). Did you mean to use the term "content attribute" here? (I agree that the spec term "DOM property" should have a better name.)

Update: I think I understand and agree with "changes to the [content] attribute are ignored once the value has been altered via the DOM property (aka IDL attribute) or by the user (e.g. by typing a new value into a field)"

**Comment by alice** [comment]
@jakearchibald 

> In this case, .style falls back to the previous value. Demo: http://jsbin.com/zumivok/edit?js,console

I'm not sure this pattern makes sense here, though, because you can think of .style as trying and failing to write to the attribute. 

A related, maybe illustrative case:

```html
<button id="el"></button>
```

```js
el.setAttribute("role", "link");  // el's role is "link"
el.accessibleNode.role = "switch";  // el's role is "switch"
el.accessibleNode.role = null;  // el's role is ... what now?
```

You could make arguments for either "link" or "button". 

 I think it boils down to: was the author trying to remove the AOM role, or remove both it *and* the ARIA role?

**Comment by cookiecrook** [comment]
@alice wrote:
> It might be simpler to follow the .value pattern and have ARIA changes do nothing once the AccessibleNode has been written to.

This is more like what I thought we had originally decided at the F2F. It seems the only addition is that content attribute now acts as the initial value setter for AOM, is that right? 

Do we need a way to determine whether the value of el.accessibleNode.role is ~reflecting the content attribute, or has been set explicitly? e.g. Is there any way for an author to know whether setting the content property will work (if AOM has never been set) or be ignored (if AOM has been set)? 

Previously, we solved this by only using the content attribute when the AOM property was null. AOM was always the source of truth until it was null, and then the markup became the source of truth. 

If the only goal for this addition is value validation, we're still going to need to spin up accessibility backing code to perform that validation, even for non-accessibility users. Otherwise, we're going to have to move a lot of the backing accessibility code out of the accessibility-only runtime context. There is a non-zero implementation and performance cost for this, and there are potential implications for privacy and security, too. 

Ultimately I want this, too, but we had agreed to wait until Phase 4 for computed values to ensure feasability/implementability of Phase 1. This change seems like it's trying to pull in little a bit of the Phase 4 scope, so I'm not convinced that's a good idea at this time.

**Comment by jakearchibald** [comment]
@alice 

> You could make arguments for either "link" or "button".

It depends on whether `null` is a valid value. With `.style`, an empty string is valid and means "unset", meaning it falls back to stylesheets etc.

If `null` (or an empty string) unsets, then yeah it's "button". If `null` is rejected since it's invalid, then it's "switch". Using an empty string to unset seems right in this case.

@cookiecrook

> Do we need a way to determine whether the value of el.accessibleNode.role is ~reflecting the content attribute, or has been set explicitly?

`.value` doesn't have a way to determine this.

**Comment by alice** [comment]
@jakearchibald 

> If `null` (or an empty string) unsets, then yeah it's "button". 

I would expect it to unset, since `null` is the value you read if it's never been set.

> If `null` is rejected since it's invalid, then it's "switch". Using an empty string to unset seems right in this case.

The explainer currently states that you can set any valid property to any value you want on `AccessibleNode`, but that [only valid values can be read _back_](https://github.com/WICG/aom/blob/master/explainer.md#validation).

This means that the browser doesn't need to do any validation until the property is being _read_, either via querying `accessibleNode` or by an accessibility API consumer.

I think it makes sense for `accessibleNode` to simply always be the source of truth, and always reflect the last thing the author _wrote_ to `accessibleNode`, or any current, valid ARIA attribute iff the author _never_ wrote to the node. So for @cookiecrook's original case that would look like:
```js
el.setAttribute("role", "foo");
el.accessibleNode.role; // null, invalid ARIA

el.setAttribute("role", "button");
el.accessibleNode.role; // "button", valid ARIA

el.accessibleNode.role = "foo"; // invalid
el.accessibleNode.role; // null, invalid accessibleNode property

el.accessibleNode.setAttribute("role", "link");
el.accessibleNode.role; // null, can't go back to ARIA any more

el.accessibleNode.role = null; // unset/clear
el.accessibleNode.role; // null, still won't fall back to ARIA

el.accessibleNode.role = el.getAttribute("role");  // you can still set this explicitly
```

**Comment by jakearchibald** [comment]
> This means that the browser doesn't need to do any validation until the property is being read

This makes it pretty different to `.style` then, so I guess it's a new pattern.

**Comment by alice** [comment]
@cookiecrook 

> Previously, we solved this by only using the content attribute when the AOM property was null. AOM was always the source of truth until it was null, and then the markup became the source of truth.

> If the only goal for this addition is value validation, we're still going to need to spin up accessibility backing code to perform that validation, even for non-accessibility users. Otherwise, we're going to have to move a lot of the backing accessibility code out of the accessibility-only runtime context. There is a non-zero implementation and performance cost for this, and there are potential implications for privacy and security, too.

I think the performance costs will be very minimal, since it doesn't actually need to spin up a whole accessibility tree, but simply perform mostly mechanical parsing (except where an IDREF needs to be dereferenced). We could (should) do it lazily when the `accessibleNode` is first queried, as well - as per the comment above.

The implementation cost is definitely not negligible, though. We were already planning to do this work in Chrome, so it's not too bad for us.

**Comment by alice** [comment]
@jakearchibald 
>This makes it pretty different to `.style` then, so I guess it's a new pattern.

Right, it's more like `.value`.

**Comment by minorninth** [comment]
How many web developers actually know how `.value` behaves, though? It seems much more obscure to me. `.style` is something that most web developers need to manipulate (or debug) a lot more often.

@cookiecrook 
> If the only goal for this addition is value validation, we're still going to need to spin up accessibility backing code to perform that validation, even for non-accessibility users. Otherwise, we're going to have to move a lot of the backing accessibility code out of the accessibility-only runtime context. There is a non-zero implementation and performance cost for this, and there are potential implications for privacy and security, too.

Agreed with Alice that doing the most mechanical parsing and validation is pretty small in the grand scheme of things, and worth pulling out of the expensive accessibility mode. It's just things like validating that aria-checked is one of "true", "false", or "mixed" and looking up the elements by id for aria-labelledby and things like that. A pretty small fraction of the code.

It sounds like web APIs are already pretty inconsistent (`.style`, `.value`), but if we wanted to be consistent with something, I still feel like `.style` would be better - in large part because Phase 4 is going to be analogous to getComputedStyle in many ways.

If we decide to just go with what makes sense and not be consistent, I think our decision should be motivated by maximizing the number of real-world use cases.

Use cases for reflecting attribute -> property
* Enables validating correct ARIA
* Enables browser feature detection

Use cases for a dirty flag: not sure?

Use cases for attribute reflecting into the property but with more of a strict precedence, where the attribute wins if set, and the property is second: I think this is the most natural way to support components. Component authors change properties only, embedders use ARIA to occasionally override when necessary.

I still think doing nothing (i.e. AOM and ARIA are kept separate) is fine because in Phase 4 we'll have computed accessibility attributes. That will enable any possible use case where an author wanted to validate or do feature detection.

**Comment by minorninth** [comment]
Thinking about this some more, the most similar to `.style` would be to reflect both ways, but if you set a property to something that's impossible to express as an attribute, we'd delete the attribute if present.

**Comment by alice** [comment]
@minorninth 

> It sounds like web APIs are already pretty inconsistent (.style, .value), but if we wanted to be consistent with something, I still feel like .style would be better - in large part because Phase 4 is going to be analogous to getComputedStyle in many ways.

> Thinking about this some more, the most similar to .style would be to reflect both ways, but if you set a property to something that's impossible to express as an attribute, we'd delete the attribute if present.

Agreed that this is the way to be consistent with `.style`. I don't think we want to do this - this will mean sprouting ARIA attributes everywhere.

> If we decide to just go with what makes sense and not be consistent, I think our decision should be motivated by maximizing the number of real-world use cases.

Also  agreed.

> Use cases for reflecting attribute -> property

> Enables validating correct ARIA
> Enables browser feature detection

> Use cases for a dirty flag: not sure?

I think the "dirty flag" is really just an implementation detail of the "reflect from attributes to properties once; once a property is set the attribute is never read again even if a new value is set to it" pattern used by value.

> Use cases for attribute reflecting into the property but with more of a strict precedence, where the attribute wins if set, and the property is second: I think this is the most natural way to support components. Component authors change properties only, embedders use ARIA to occasionally override when necessary.

So this is the style where authors would need to remove ARIA attributes in order to have their `accessibleNode` properties apply?

**Comment by cookiecrook** [comment]
@minorninth wrote:

> Agreed with Alice that doing the most mechanical parsing and validation is pretty small in the grand scheme of things, and worth pulling out of the expensive accessibility mode. It's just things like validating that aria-checked is one of "true", "false", or "mixed" and looking up the elements by id for aria-labelledby and things like that. A pretty small fraction of the code.

I think there may be cases where the mechanical value validation could be misleading. For example: `<div role="cell">Not in a table</div>` The value validation proposal would return `cell` for the `accessibleNode.role`, but the computedRole would be something else: probably a generic with no exact ARIA role match, because the cell is orphaned from a table.

In Phase 4, the plan is to return the actual computedRole value, but it seems the new proposal is to use simple token-matched validation in Phase 1. Is that correct? Wouldn't this mean some Phase 1 values were incompatible with the later return values planned for Phase 4? Do we care that these will change?

**Comment by alice** [comment]
Overall, I'm still reasonably convinced that a hybrid of `.value` and `.style` patterns is what we want:

**Like `.value:`**
-  attributes reflect to properties but not vice versa (more on this below);
- "dirty state" pattern, i.e. once property is set, attribute is _completely disregarded including any changes_ (more on this below);

**Like `.style`:**
- do validation on property setting (more on this below).

### Reflection

I honestly don't have a super strong argument here. My thinking goes more or less like:
- Reflecting ARIA to `accessibleNode` will improve ARIA's ergonomics and "explain" ARIA.
- Reflecting `accessibleNode` to ARIA (à la `.style`) will sprout attributes everywhere and be annoying. Moreover, we *can't* reflect everything, so it'll be inconsistent, with some attributes being magically removed instead of updating.

I think this plus my preferred option for precedence more or less get us to the "dirty state" pattern.

### Precedence

I think there are three main options:
1. ARIA always takes precedence. Authors must remove ARIA attributes to have `accessibleNode` properties win.
2. `accessibleNode` properties take precedence. Custom Element authors _should_ check for ARIA. Authors must use `accessibleNode` to be completely sure that their properties apply. 
3. Last one wins. There is no way to reliably detect what has applied until phase 4 (see below for my qualms about this).

My strong preference has changed since [earlier](https://github.com/WICG/aom/issues/60#issuecomment-287183115), and I now think (2) is the way to go:
- Custom Element authors _should_ support ARIA, but we don't need to force them to do the right thing. Also, if a Custom Element never writes to a given `accessibleNode` property, ARIA will still work; it's not like Custom Element authors would need to write code to support all potential ARIA properties, only the ones they have overridden.
- In any case, this is not a Custom Element specific API, so there's nothing preventing page authors from simply writing to a Custom Element's `accessibleNode` to ensure they have the last say:
```html
<custom-checkbox id="chips">Chips</custom-checkbox>
```
```js
chips.accessibleNode.label = "Fries";
```
- We can see this as "explaining" ARIA: ARIA sets a property on the `accessibleNode`, which is used as an input to the computed properties of the corresponding node in the computed accessibility tree;
- This means that implementation-wise we can also abstract ARIA away from our accessibility tree implementations.

Yes, we do now have to educate people about why their ARIA might not apply, but I can't see a situation in which *existing* ARIA-based code will come into conflict with *new* `accessibleNode` based code and "break" (maybe others can?) 

### Validation

@minorninth wrote:
> I still think doing nothing (i.e. AOM and ARIA are kept separate) is fine because in Phase 4 we'll have computed accessibility attributes. That will enable any possible use case where an author wanted to validate or do feature detection.

(@asurkov raised something similar on the mailing list, and @robdodson also mentioned this in conversation.)

I'm not sure the above is true. We don't yet know what Phase 4 will look like, but we know that it will entail spinning up an accessibility tree, at least partially. 

- This seems like a huge cost just to do mechanical validation on properties which will, for the vast majority of users, never be read by assistive technology. Can you imagine authors incurring that performance impact *for all users* just to check their semantics?
- It also implies that the phase 4 API may well be asynchronous, which again makes this very unwieldy for authors.

Here are some [CSS Typed OM](https://www.w3.org/TR/css-typed-om-1/) inspired ideas for validation without waiting for phase 4 (note that these are by no means intended to be mutually exclusive; we may want none, some or all of these):
- Provide an explicit hook for feature detection, along the lines of [CSS.supports()](https://developer.mozilla.org/en/docs/Web/API/CSS/supports)
```js
if (AOM.supports("role", "switch"))  // TODO: bikeshed where this method should live
    el.accessibleNode.role = "switch";
else
    el.accessibleNode.role = "button";
```
- Provide enum constants for everything - apparently CSS OM will have this for the keyword types eventually.
```js
if ("ROLE_SWITCH" in AOM)  // TODO: bikeshed where enums should live
    el.accessibleNode.role = AOM.ROLE_SWITCH;  // TODO: bikeshed enum structure
else
    el.accessibleNode.role = AOM.ROLE_BUTTON;
```
- Have an explicit setter method which throws an exception if an invalid value is used (would definitely want a way to check validity without a try/catch):
```js
el.accessibleNode.set("role", "banana");  // dies
// or
el.accessibleNode.setRole("banana");  // dies

// explicit getter?
el.accessibleNode.role();  // null, also this line is never reached because it already died

// ES6 getter with no setter?
el.accessibleNode.role; // as above

// Generic getter?
el.accessibleNode.get("role");  // same again
```

**Comment by alice** [comment]
@cookiecrook 
> I think there may be cases where the mechanical value validation could be misleading. For example: `<div role="cell">Not in a table</div>` The value validation proposal would return `cell` for the `accessibleNode.role`, but the computedRole would be something else: probably a generic with no exact ARIA role match, because the cell is orphaned from a table.

This is a good point, but I think it's ok. I think it's like the difference between
```js
myEl.style.color = "cumquat";  // not a valid value
```
and
```css
body * {
  color: red !important;
}
```
```js
myEl.style.color = "blue";  // not !important enough to apply, but valid
```
We can't know until we compute the tree whether the role will "apply", but we can tell the author their attempt to apply the role was well-formed.

> In Phase 4, the plan is to return the actual computedRole value, but it seems the proposal is to use simple token-matched validation in Phase 1. Is that correct? Wouldn't this mean some Phase 1 values were incompatible with the later return values planned for Phase 4? Do we care that these will change?

Yes, that seems right to me.

**Comment by robdodson** [comment]
> My strong preference has changed since earlier, and I now think (2) is the way to go

+1

> Provide an explicit hook for feature detection, along the lines of CSS.supports()

+1

I'm a fan of the `CSS.supports` approach proposed by @alice, as opposed to the magic validation where setting a property can result in `null`. My thinking is `accessibleNode` seems like a POJO, and many folks will assume that when they set `accessibleNode.role = 'switch'` that all they've done is set a property.

A very likely scenario I can see playing out is someone does `accessibleNode.role = 'switch'` and it works on all the browsers that they test. Then they get a bug saying the role is `null` in Browser X that they forgot to test. The developer wonders "How could it be null? I can see in the code I set it to 'switch' and it works in browsers A, B, and C. Is this AOM thing just buggy in Browser X?" Now they have to go on a wild goose hunt to learn about the magical null validation behavior.

I'd much rather `role` be whatever I set it to (right or wrong) and there's an explicit API for checking if my role is actually supported.

**Comment by robdodson** [comment]
Follow up after discussing this with some other devrel folks and Custom Element authors.

I think I'd prefer AOM win versus ARIA because that's consistent with how attributes in custom elements work today. For example `<x-btn disabled role="switch">`. When my element boots up I already have to check if there's a disabled attribute, otherwise I'd incorrectly set a default value of `this.disabled = false`. Similarly I should check if the user has applied a role before applying my default one.

Ideally I'd like whatever the user sets as the `role` attribute to reflect to `accessibleNode.role`. That way I could do something like this in my Custom Element:

```
if (!this.accessibleNode.role) {
  this.accessibleNode.role = "button"
}
```

This is also consistent with how `value` works today.

I'm also against the validation `null` behavior because it's a bit too magical. That `style` does the empty string thing doesn't feel like a good behavior to emulate. I've spent an afternoon scratching my head about the empty string thing before I realized I had a typo. Instead I'd much rather `accessibleNode` either throw an error or log a warning if I set an unsupported `role`. But I'd like it to leave whatever I set in place.

This, I hope, follows the KISS principle and avoids the user having to retain a deep understanding of any special rules.

**Comment by robdodson** [comment]
As @minorninth pointed out, throwing an error would break graceful degradation in older browsers so I guess that's out :)

I still vote for a warning though.

**Comment by minorninth** [comment]
Here's a summary of what I'm hearing from everyone:

*Precedence*: There are pros and cons to each, but there's more support for AOM properties taking precedence over ARIA attributes. Neither solution is perfect.

*Reflection*: I'm not hearing any strong objections to the idea that ARIA attributes are reflected into AOM properties, in one direction.

*Validation*: Several people have concerns about whether this is the right thing to do, or how it should be implemented if so. There are several ways we could support the use cases of validation and feature detection without validating in the getter.

I propose the following;

* ARIA attributes are reflected into AOM properties (one-way), but once an AOM property is set, that takes precedence from then on (a dirty bit). Like .value.
* AOM properties override ARIA attributes. This is already implied by the sentence above, just making it clear.
* No validation in the getter. If you set element.accessibleNode.role = 'beyonce', then you get back 'beyonce'.
* In a separate phase (possibly before what's currently phase 4), we add something like AccessibleNode.supports() to provide authors a way to do feature detection or validation, without confusing it with reading the computed value. We also optionally encourage user agents to fire warnings if setting a property to an invalid value. Neither of these changes break any code written for phase 1.

**Comment by robdodson** [comment]
I think this all aligns with @alice's original proposal from https://github.com/WICG/aom/issues/60#issuecomment-288365855

So yes +1 from me :D

**Comment by alice** [comment]
Yup, this is basically what I was proposing earlier (modulo validation), so +1 from me.

**Comment by fleizach** [comment]
I think we can do basic role validation in WebCore, but doing computed roles would be much more challenging, since we'd need to create a tree to do so.

So I think

```html
<div role="cell">Not in a table</div>
```
returns `div.accessibleNode.role == "cell"`

but

```html
<div role="blah">
```
returns `div.accessibleNode.role == null`

then it looks good

**Comment by minorninth** [comment]
The main difference from Alice's earlier comment is that we'll agree to set aside validation until later. In Phase 1 we'll do no validation.

@cookiecrook and @asurkov - how does this sound?

@fleizach eventuall we'll all need to figure out a way to provide computed attributes somewhat efficiently. I think there are ways we could do that efficiently enough (we have some code now that creates a tree on the stack and does only minimal computation) but we can worry about that later.

**Comment by alice** [comment]
It seems like we have consensus from Chrome folks on reflection and precedence. I've filed a separate issue #70 on validation, since that was introduced late in the piece and I'm not sure we fully explored all the options, and I don't want to mistake exhaustion for consensus when this is hopefully going to be the API for a long time.

**Comment by asurkov** [comment]
* ARIA to AOM reflection makes sense with me when it's part of computed value. It seems less valuable if separated from native markup semantics. It may also add some implementation overhead by keeping separated platform's and AOM properties computation.

* Thumb up for AOM overriding ARIA.

* I'm also good to drop validation and feature detection from phase 1

**Comment by alice** [comment]
Thanks for the comments @asurkov! Seems like we have full consensus around AOM overriding ARIA.

Regarding reflection, I'm not sure I fully understand your comment, perhaps you or others could help clarify: 

> ARIA to AOM reflection makes sense with me when it's part of computed value.

This seems unavoidable, since the computed value will take all potential sources into account in any case, right?

> It seems less valuable if separated from native markup semantics.

In other words, being able to reflect the ARIA role but not the computed role from native markup (e.g. `<button>` -> `accessibleNode.role === "button"`)?

I think reflecting ARIA but not native semantics makes sense - if you think of AOM as "another way to use ARIA semantics". So for an element with a `role` attribute, you can choose between 

```js
let authorRole = el.getAttribute("role");
```

and


```js
let authorRole = el.accessibleNode.role;
```

to know what role has been *set* on the element.

Then, there's no need to ever use `getAttribute` in practice, since `accessibleNode.role` will reflect the attribute or the AOM property which has overridden it.

In terms of what *author value* this has, I imagine it would mostly be used by library widget or custom element authors to detect what semantics have been set by authors using their component, to avoid overriding it. 

> It may also add some implementation overhead by keeping separated platform's and AOM properties computation.

I don't quite follow what's being referred to here.

I think at this point we (meaning @minorninth, me and maybe @cookiecrook) imagine keeping the computed properties fairly distinct from the author-set properties, if that's what you're referring to? So it would be closer to `window.getComputedStyle(el)` than `el.style`, in other words. @cookiecrook proposed something like (I forget the exact syntax, sorry) `el.accessibleNode.getComputed("role", "checked", "disabled");` to get a snapshot of a subset of attributes.

**Comment by domenic** [comment]
I'm sorry to say I haven't had time to keep up with this thread, but scrolling to the bottom, "AOM overriding ARIA" seems unfortunate from an explaining the platform/custom elements point of view. I had always hoped that as a custom elements author I could use AOM to set up the "intrinsic" semantics of my element, not reflected into the DOM---but then authors could override the semantics if necessary using ARIA attributes. So I was hoping ARIA would override AOM, just like ARIA overrides the browser's built-in a11y roles/states/properties for built-in elements.

The precedence argument in https://github.com/WICG/aom/issues/60#issuecomment-288365855 is reasonable, so I am OK being sad about this not going the way I prefer. It basically means we'll still not have a mechanism for custom elements to act like built-in elements, but I understand that's not the priority of this API.

Separate from precedence issues, I'd **strongly** urge you not to copy the `.value` pattern with its dirty bit. That is one of the most confusing patterns we have on the web platform, tripping up authors daily (check e.g. StackOverflow questions for evidence), and I hope we never repeat that mistake in further APIs.

**Comment by alice** [comment]
We can't really separate precedence issues from the `.value` pattern. We're not copying it to be consistent, we're copying it because having ARIA reflect to AOM, AOM not reflecting to ARIA and having AOM take precedence are all separately things we want, and more or less _amount_ to the `.value` pattern.

**Comment by domenic** [comment]
OK. I'm saying that you should not want something that more or less amounts to the `.value` pattern, because of the mass of historical evidence that points to this being a nightmare for developers.

I'd strongly suggest rethinking one of those requirements, if they inevitably lead to this historically-bad pattern.

In my mind of the three you listed, only "AOM not reflecting to ARIA" is crucial; the other two should be flexible if the consequences of them are bad, which I think they are in this case.

**Comment by alice** [comment]
Could you explain why that one is crucial?

**Comment by domenic** [comment]
If AOM reflects to ARIA, then that ties AOM's power to be exactly as much as ARIA already provides, which obviates most of its benefits. (Either that, or we invent some new meaning of reflection different from the rest of the IDL attribute <-> content attribute reflection on the rest of the platform, which is also not great.)

**Comment by alice** [comment]
The same argument kind of applies for having AOM take precedence, doesn't it?

Imagine you had a custom element, `<my-element>`, which an author had set `aria-activedescendant` on as a starting value. The custom element wants to change that to something within a shadow root, so it uses AOM. Now it has to remove the ARIA attribute to ensure the AOM property applies, even though there's no way to even express its current state in ARIA.

Even for things you can express in ARIA, you would have to either remove the ARIA attribute or else constantly be updating it to match the AOM value (i.e. implementing reflection yourself).

**Comment by domenic** [comment]
I don't really understand how that bears on the idea that reflection already has a defined processing model, sorry.

In your "imagine" scenario, what semantics are we using? ARIA has priority, or AOM has priority, or AOM reflects ARIA and therefore there's no question of priority since they're always identical?

**Comment by alice** [comment]
I'm saying imagine that ARIA takes priority instead of AOM, but the other things remain the same.

**Comment by domenic** [comment]
OK. So I agree that if ARIA takes priority, then ARIA takes priority ("now it has to remove the ARIA attribute"). I'm still not sure how this relates to the overall discussion.

**Comment by alice** [comment]
So in order to use AOM, you will have to constantly be touching ARIA attributes as well.

**Comment by domenic** [comment]
I don't think that's true. You can use AOM just fine, and let ARIA attributes be used rarely as an override by page authors. If page authors have chosen to give your element a specific role/state/property, you shouldn't second-guess them.

This is just like today you can use `<a>` or `<button>` just fine, and let ARIA attributes be used rarely as an override by page authors. We don't override author choices for those elements, I don't think.

**Comment by alice** [comment]
States, by definition, will change. Things like `aria-activedescendant`, by design, will change. These things are intended to be dynamic. It's not a matter of overriding an author's intent, in many cases, but in recognising that intent was a _starting state_.

**Comment by domenic** [comment]
OK, that makes sense. Then people can set those dynamically. I'm still not understanding why this necessitates a `.value`-like design.

**Comment by alice** [comment]
You mean Custom Element authors can set them dynamically? They would have to reflect back to ARIA themselves, or remove it.

**Comment by domenic** [comment]
Why would they do that? Can you explain to me the motivations of each party here (the custom element author and the page author)?

Maybe this is the scenario you are envisioning?

- Custom element author uses AOM
- Author uses ARIA
- But author didn't mean to override the custom element author, so the result is unexpected?

I have thoughts on this, but want to confirm I'm understanding before going into them.

**Comment by alice** [comment]
The scenario would be something like:

- Author sets `aria-checked` on a custom element
- User takes some kind of action causing the checked state to toggle
- Custom element code wants to update the accessible checked state, but needs to remove or update the ARIA attribute first.

**Comment by domenic** [comment]
Ah, OK. Then my thoughts are that the author should not be setting aria-checked on the custom element, just like they wouldn't be setting it on a `<input type="checkbox">`, unless they are sure they want to override the custom element/checkbox's accessible checked state.

**Comment by alice** [comment]
Hm, ok then.

Would you then have ARIA and AOM be completely separate systems, other than ARIA taking precedence? i.e. no reflection at all?

**Comment by domenic** [comment]
I think that makes sense to me, but now that we're outside of a specific scenario I have to be more cautious about stating my position as a "good" one, since I didn't get a chance to read through this whole thread and see all the motivations involved. I feel that `.value`-like is quite bad, but maybe there are ways of making them interact between `.value`-like and no-interaction.

I can't really see the downsides of no interaction from where I'm standing, though.

**Comment by alice** [comment]
My instinct at this point is that we have either `.value`-like, or no interaction between them other than ARIA taking precedence, since full two-way reflection isn't an option.

**Comment by alice** [comment]
I think the downside of "no reflection" is that these things share a vocabulary, so it would be reasonable to expect that they were different expressions of the same thing. I think it might be confusing and annoying to authors that there is still no programmatic way beyond `setAttribute()` and `getAttribute()` to set ARIA, which has the final word in semantics.

@robdodson makes this point indirectly here: https://github.com/WICG/aom/issues/60#issuecomment-287201660

**Comment by minorninth** [comment]
> The scenario would be something like:
> 
> Author sets aria-checked on a custom element
> User takes some kind of action causing the checked state to toggle
> Custom element code wants to update the accessible checked state, but needs to remove or update the ARIA attribute first.

Alice, I also don't find this example very compelling, because it's not clear why you wouldn't want the page author to win. In general we should want custom elements to work well by default, and encapsulate all of their internal state as possible, but we also want page authors to be able to override them as much as possible, too.

Also, I think it's different from the argument Rob was trying to make earlier.

> I think I'd prefer AOM win versus ARIA because that's consistent with how attributes in custom elements work today. For example <x-btn disabled role="switch">. When my element boots up I already have to check if there's a disabled attribute, otherwise I'd incorrectly set a default value of this.disabled = false. Similarly I should check if the user has applied a role before applying my default one.

I think Rob's argument was that the custom element author may want to change its behavior, such as its initial state, based on the attributes provided by the page author.

However, I'm not sure that they're the same. When the author sets the disabled attribute on x-btn, that's part of x-btn's public API. The author specifically wants the x-btn to change its visual appearance and stop responding to input events. When the author says role="switch", though, they're not really asking the x-btn to change its appearance or semantics.

So I guess that's an argument in favor of ARIA taking precedence over AOM.

It's a bit unfortunate because I did like the idea of reflection. I liked the idea of reflecting ARIA attributes into accessibleNode properties, to show their correspondence and provide some nice discoverability and exploration.

However, purely from the perspective of intended use cases and not for testing, validation, feature detection, and exploration - I suspect "ARIA wins" is more *useful*, in that it balances the power reasonably well between component authors and page authors.

I think part of the problem here is that without the last phase, the ergonomics are poor because there's no easy way to figure out what wins. It's like having "style" without getComputedStyle.

Here's an idea: what if we added support for getting computed properties back to the spec now - back in Phase 1 or 2 - but we specifically exclude properties that depend on things like layout.

Role seems safe, for example.

So what I'm thinking is that we define exactly what it'd look like to expose the computed role in the AOM, along with things like the computed "checked" state and many others. Then we defer things like exposing the computed name and parents and children because there are lots of issues around those.

I wonder if we did this and we had clear, nice-looking semantics for querying the computed role, then it might be fine if there was no reflection.

To be very explicit, we could use "source" as Alex suggested:

```
element.accessibleNode.source.role = "button";
element.accessibleNode.computed.role;  // returns "button"

element.setAttribute("role", "switch");
element.accessibleNode.computed.role;  // returns "switch"
```

**Comment by alice** [comment]
I think layout could affect computed role.

For example:

```html
<div id="parent" role="tree">
   <div id="child" role="checkbox">Checkbox can't be a direct child of a tree, this is just text</div>
</div>
```

What should `child.accessibleNode.computed.role` be?

**Comment by alice** [comment]
> it's not clear why you wouldn't want the page author to win. 

In this case, my argument was that because this is a dynamic state, the fact that the state has changed means that the page author-provided value is now out of date.

However, I agree that this scenario shouldn't really happen in practice - custom elements should provide a non-ARIA API for setting that type of state.

> I think it's different from the argument Rob was trying to make earlier.

I actually meant [this comment](https://github.com/WICG/aom/issues/60#issuecomment-287201660) (I think we have timing issues with deep-linking and loading causing the page to scroll; it would be nice if GitHub would provide a strong visual indication of the deep linked comment!):

> > Reflecting the ARIA attribute to the property sounds good to me. In fact I would _really_ want this. 

> However, purely from the perspective of intended use cases and not for testing, validation, feature detection, and exploration - I suspect "ARIA wins" is more useful, in that it balances the power reasonably well between component authors and page authors.

Let's not forget that page authors can also use AOM. I think it's more a question of static, load-time semantics vs dynamic updates.

For example, there is a developing responsive pattern where a particular role is specified in markup, and then overridden in script when the script loads and provides fuller interactivity. There is a good demonstration in [howto-tabs](https://github.com/GoogleChrome/howto-components/tree/master/elements/howto-tabs):

- The [demo page](https://github.com/GoogleChrome/howto-components/blob/master/elements/howto-tabs/demo.html) hard-codes a role of `heading` for tabs and a role of `section` for tabpanels, meaning that if the script never loads or executes, the content can be understood as a series of sections and headings.
- The [`connectedCallback`](https://github.com/GoogleChrome/howto-components/blob/master/elements/howto-tabs/howto-tabs.js#L48) hook in the custom element code rewrites the roles to `tab` and `tabpanel` to reflect the interactive version's semantics.

In this case, it's not a matter of "overriding" the page author's wishes, but that the `tab`/`tabpanel` roles only make sense when the script has loaded.

Obviously, this could be implemented by removing or rewriting the ARIA, but the outcome is still the same.

> I think part of the problem here is that without the last phase, the ergonomics are poor because there's no easy way to figure out what wins. It's like having "style" without getComputedStyle.

Modulo strong native semantics (which I think we should leave aside) and layout-dependent issues like in the above comment, I think it's still "easy", just clumsy:

```js
let role = el.hasAttribute('role') ? el.getAttribute('role') : el.accessibleNode.source.role;
```

I'm still opposed to using computed values for any run-time logic.

**Comment by alice** [comment]
What if it was last one wins, but `accessibleNode` (or `accessibleNode.source`) always reflects the current value?

i.e.

```html
<div id="composeBtn" role="button">Compose<div>
```

```js
composeBtn.accessibleNode.role; // button
composeBtn.accessibleNode.role = "switch";
composeBtn.accessibleNode.role; // switch
composeBtn.setAttribute("role", "checkbox")
composeBtn.accessibleNode.role;  // checkbox
```

**Comment by domenic** [comment]
That seems like it might work. However, I'm not sure what the underlying model is. Is it something like this?

- Every accessible node has a [[role]] internal slot
- Every time the role="" content attribute changes, we override [[role]] to contain the attribute's new value. (Or, only if it's a valid role?)
- The `accessibleNode.role` setter also overrides [[role]] to equal the given value. (Or, only if it's a valid role?)
- The `accessibleNode.role` getter returns [[role]].
- Through some process un-specced here, [[role]] feeds into the a11y machinery to determine the "computed role" exposed to the AT user.

(I recognize the validity question is being discussed in a separate thread, and we don't have to get into it here, but I wanted to highlight that a model like this gives you two separate places you need to decide the validity question, not just one.)

I think this model is OK, but it does have some drawbacks, such as providing two ways to do the same thing. It also might be a bit confusing for authors that the content attribute affects the IDL attribute, but not in the usual two-way manner that we see with content attribute <-> IDL attribute reflection.

Plus of course it loses the ability to have the two-tier structure of underlying-role and role-provided-by-DOM-attributes, which I so desperately want so I can emulate native element behavior with custom elements. But I understand that's not this API's priority.

Overall I think it's more powerful to have the two different APIs operate on different levels. Your suggestion makes them operate on the same level, which means we aren't exposing new capabilities so much as providing a new syntax for doing the same thing. But, this new syntax has benefits like being able to refer to other elements without idrefs, and avoid sprouting attributes, so maybe that's enough of a benefit. I can understand if the extra power necessary to emulate `<input type="checkbox">` behavior is considered out of scope.

**Comment by domenic** [comment]
To be a bit more concrete, one of the downsides of the above model is the following code:

```js
composeBtn.accessibleNode.role; // button
composeBtn.accessibleNode.role = "switch";
composeBtn.accessibleNode.role; // switch

composeBtn.getAttribute("role"); // button <--- maybe unexpected

composeBtn.removeAttribute("role");
composeBtn.accessibleNode.role; // empty string <-- pretty unexpected IMO; what happened to "switch"?
```

**Comment by alice** [comment]
@domenic Thanks for writing that example out. I hadn't considered the removing attribute case - I agree that's weird (although I think every option we can come up with has some kind of weirdness; it's a matter of which type of weirdness we want to live with).

Regarding your notes on the underlying model - that seems right.

There were a few parts of your comment I didn't quite follow: 

Regarding validity, I think we have the same issue if ARIA doesn't reflect but still takes precedence; does invalid ARIA still "override" a valid AOM property? i.e. `<my-element id="el" role="🍌">`, `el.accessibleNode.role = 'button';` - what would you expect the computed role here to be? Presumably `"none"` or `null`, but I'm not clear on why this constitutes fewer places to decide the validity question.

I don't quite follow that you say you don't want two ways to do the same thing, but you don't mind having two totally separate systems with the same vocabulary affecting the same outcome - that seems like two ways to do the same thing to me:

```html
<my-element id="el">
```

```js
el.setAttribute('role', 'switch');
// or, and equally valid and having the same effect on the computed accessibility tree
el.accessibleNode.role = 'switch';
```

I also don't follow this:  
> I can understand if the extra power necessary to emulate `<input type="checkbox">` behavior is considered out of scope.

What specific power were you referring to?

**Comment by domenic** [comment]
> Regarding validity, I think we have the same issue if ARIA doesn't reflect but still takes precedence; does invalid ARIA still "override" a valid AOM property?

Yeah, although I wouldn't call it the same issue, this is indeed an issue to be discussed. I'd suggest going with whatever the behavior is right now for e.g. `<input type="checkbox" role="banana">`.

> I don't quite follow that you say you don't want two ways to do the same thing, but you don't mind having two totally separate systems with the same vocabulary affecting the same outcome - that seems like two ways to do the same thing to me:

I'm confused by your confusion. You stated the difference quite well: it's "two ways to do the same thing" (operating on a single underlying system) vs. "two separate systems affecting the same outcome". Notably those code examples don't do the same thing, because you can vary each independently to achieve useful effects (again, like those of native elements).

> What specific power were you referring to?

The ability to have a set of underlying semantics, not controlled by content attributes, but able to be overridden by them. (And then fall back to the underlying semantics if the content attributes are removed.)

That's exactly what built-in elements have today, and what you would get with an ARIA-layers-on-top-of-AOM model, but could not get with the last-one-wins model.

**Comment by minorninth** [comment]
What I like the most about what's in the explainer now is reflection. I think that's great for discoverability. I also liked the limited amount of validation / type-checking we had originally, I thought that was a great fit.

It doesn't look like we're going to get validation. Mozilla isn't supportive, and Apple says it will be more work for dubious benefit. Also I don't think any of the core use cases depend on it.

Out of the approaches discussed, I don't really see how one of them is "two ways to do the same thing" but the other is not. With any of the approaches we're talking about, there's a DOM attribute and an AccessibleNode property, they can have different values, the property is richer, and they both affect one underlying computed property that's communicated to assistive technology. What's different is how the attribute and property relate to one another, but in any of the approaches they're distinct.

Consider what happens with the "dirty value flag" model when you consider not only what the AccessibleNode property returns, but the actual computed value too.

```
element = document.createElement('a');
element.accessibleNode.role;  // returns null
// Step 1. Computed role is "link"

element.setAttribute('role', 'button');
element.accessibleNode.role;  // returns "button"
// Step 2. Computed role is "button"

element.accessibleNode.role = 'checkbox';
element.accessibleNode.role;  // returns "checkbox"
// Step 3. Computed role is "checkbox"

element.accessibleNode.role = null;
element.accessibleNode.role;  // returns null
// Step 4. Computed role is now "link" even though the ARIA role is "button"
```

This just seems really bad to me, because from the perspective of the author, the state of the world at step 2 and step 4 are indistinguishable. Not only is there a magic hidden flag that you can't see, you can actually get into a state where it's impossible to predict the outcome by examining all of the properties and attributes you can access from JavaScript.

Correct me if I'm wrong, but that's worse than the input "value" attribute, and I think the key difference is that "value" isn't nullable.

If you set the value attribute, that's reflected in the value property. So far so good.

If you then change the value property, the value property now takes precedence. Changing the value of the attribute no longer matters. That's maybe a bit confusing, but the value property still reflects reality. You can't set it to null. In fact, it validates - so if you try to set it to something invalid, you'll get back something slightly different.

The bottom line is, you can always know the value of an input element by checking the value property.

As currently written, the same isn't true for AOM, and that seems like worse than other approaches that have different tradeoffs.

At first I tried to convince myself there was an approach where AOM properties could override ARIA attributes, but they were clearable, as a way to avoid that problem. But then you can't have reflection, *unless you also have validation*.

If we have validation, then *setting* an AOM property to null means to clear any extra semantics from the AccessibleNode, and revert back to as if it hadn't been overridden, and *getting* an AOM property always means get me the latest validated value from either the property, or from the corresponding ARIA attribute. However, making this work means buying into validation, which I don't think we can do.

I think that unfortunately what that means is that we have to give up reflection.

If we give up reflection, then either AOM properties or ARIA attributes can take precedence and both are equally consistent and don't lead to any weird cases. I think I've always leaned towards ARIA attributes taking precedence, all other things being equal, and Domenic's arguments above reinforce that. I was okay with AOM taking precedence if we could make reflection work, but now I think I'm convinced that that's going to be inferior.

To summarize, I think I'm going back to ARIA attributes overriding AOM properties, with no reflection or validation either way. It's not perfect but it's consistent and not at all surprising once you understand it, and in a later phase we'll provide a way to get computed values.

**Comment by cookiecrook** [comment]
@minornight wrote:
> To summarize, I think I'm going back to ARIA attributes overriding AOM properties, 

I've been opposed to ARIA overriding AOM all along. I think @asurkov has been too. I may have missed (or forgotten) a reason listed in the giant thread above, but IIRC, the core needs were 1) author discoverability, and 2) a novice/intermediate author's ability to override.

**1. Author discoverability.** I understand this case, but I think improvements to dev tools are the only good solution to this problem. Don't design an API based on the convenience of using it with the current tools. Design the tools to make it convenient to use the best API you can envision. Regarding tools: WebKit and Edge have engine-computed values in the dev tools, and @alice has an [outstanding Chromium patch to allow something similar](https://bugs.chromium.org/p/chromium/issues/detail?id=390254). Firefox has an inspector with computed values but it seems targeted at implementors (by Alex, for Alex) rather than web developers. To my knowledge, all the add-on accessibility extensions (including the shipping Chrome accessibility extension) vend accessibility property info based on assumptions of DOM parsing rather than computed values in the engine. @minorninth @alice @asurkov Is that accurate and fair? If so, do you see improving the dev tools as an adequate solution to this problem?

**2) A novice/intermediate author's ability to override AOM by setting an ARIA attr.** If I understand this argument correctly (I assume I'm missing something), the story is that someone might need to override a framework that uses AOM, but they don't know how to modify the AOM? If that's an accurate portrayal, I don't give this argument any credence. (Analogy: Most CSS novices learn about `!important` but it's better to learn and rely on the nuances of the cascade.) Any author in the situation should learn what is necessary to override AOM with AOM.

Like ARIA overrides native role semantics and most properties, I hoped AOM would override ARIA:

- AOM overrides ARIA.
- ARIA overrides host language (e.g. HTML). [Implemented today.]

Also:
- Nulling an AOM value falls back to ARIA. 
- Nulling an ARIA value (removeAttribute) falls back to the host language. [Implemented today.]

_It feels really backward to me to have a JavaScript interface that defers to (or can be overridden by) markup/DOM at some times (role and aria-* attrs) but not at other times (tagname, alt attr, required attr)._ 

> To summarize, I think I'm going back to […snip…] no reflection or validation either way. 

I'm also back around to "no reflection" because, as you seem to imply below, the real reason for reflection is to get something closer computed values. If we can't get real computed values yet, let's wait until we can, and not complicate the path to get there.

> It's not perfect but it's consistent and not at all surprising once you understand it, and in a later phase we'll provide a way to get computed values.

+1

**Comment by cookiecrook** [comment]
In summary, my vote is: AOM overrides ARIA with no reflection.

**Comment by minorninth** [comment]
@cookiecrook my only disagreement is really with your characterization of reason #2 as "A novice/intermediate author's ability to override AOM by setting an ARIA attr". I don't think it has to do with being novice, it's the idea that setting an ARIA attribute would be a natural way to override a custom control that has native semantics.

So if somebody implemented <custom-slider>, and I wanted to use it as a scroll bar, I'd love to just write:

    <custom-slider role="scrollbar">

If AOM overrides ARIA, it means the author of the custom element would need to check for the attribute and maybe not set the AccessibleNode property if it's already set. I guess that isn't so bad, it'd just be nice if they didn't have to remember to do that. It'd be a best practice.

Anyway, that's just clarifying one argument, but I can live with it either way.

Alice, how are you feeling about this now?

**Comment by domenic** [comment]
> If AOM overrides ARIA, it means the author of the custom element would need to check for the attribute and maybe not set the AccessibleNode property if it's already set. I guess that isn't so bad, it'd just be nice if they didn't have to remember to do that. It'd be a best practice.

It's a little worse than that. They'd need to listen for all changes to every ARIA attribute that applies, and if the attribute is added or changed, update the corresponding property via AOM. And if the attribute is removed, then restore the control's default role/property/state.

This is a significant amount of boilerplate to write, and easy to get wrong, very similar to the easy-to-get-wrong boilerplate for combining custom elements and ARIA today.

So yeah, it's workable, but it's definitely worse for custom control authors.

**Comment by alice** [comment]
I can live with more or less any of the options discussed here.

As Dominic put it, I'm disappointed not to have reflection; in my case because I'd like AOM to feel more like a system underpinning ARIA than a completely different system.

Dominic and I chatted today and agreed that basically implementing James' [proposal](https://github.com/WICG/aom/issues/60#issuecomment-298871352) (which seems to be the common ground among implementers), and getting it in front of developers, is probably going to be the quickest way to shake out how painful or otherwise this will be for developers.

**Comment by robdodson** [comment]
> Dominic and I chatted today and agreed that basically implementing James' proposal (which seems to be the common ground among implementers), and getting it in front of developers, is probably going to be the quickest way to shake out how painful or otherwise this will be for developers.

I'd be happy to try the various suggested approaches and give feedback on them. I've been working with some teammates to create a set of Custom Elements which implement all of the [ARIA Authoring Practices Guide](https://www.w3.org/TR/wai-aria-practices-1.1/) recommendations. I think we'd be able to point out any roadblocks that the larger developer community would encounter.

**Comment by surma** [comment]
Just caught up on this! I think I agree with the current decision. I think. I can’t quite tell. I am gonna state my thoughts and y’all can feel free to ignore them. 

I tried to form an informed opinion on what exactly “ARIA reflects to AOM” should entail and I am a bit confused, esp. after reading [@domenic’s example](https://github.com/WICG/aom/issues/60#issuecomment-298047752):

```html
<el role="button"></el>
<script>
el.getAttribute('role'); // 'button'
el.accessibleNode.role; // 'button'
el.accessibleNode.role = 'list';
el.getAttribute('role'); // 'button'
el.accessibleNode.role; // 'list'
el.setAttribute('role', 'heading');
el.accessibleNode.role; // (1) ???
el.removeAttribute('role')
el.accessibleNode.role; // (2) ???
</script>
```

_My_ expectation is for (1) to return `heading`, and (2) to  return "" or whatever the appropriate value for “unset” is. So when we say “ARIA reflects to AOM”, I’d expect that to always happen, not only once at the start. So I actually disagree that @domenic’s example code shows unexpected behavior. 

However, taking a look at [how `<input>` and `.value` behave](https://jsbin.com/qeculuy/edit?js,console), both (1) and (2) should return `list`. I am not sure what to make of that, I am honestly fine with either at this point. 

I do however think that ARIA reflecting to AOM and having AOM be the source of truth (_not_ requiring developers to remove ARIA attributes first to be able to start using  AOM) is the right decision in terms of ergonomics. I don’t think we have anything on the platform that behaves this way.

As @robdodson said, we’re happy to do write some practical examples using different approaches to AOM and see how it affects developer ergonomnics.

**Comment by minorninth** [comment]
Based on Alex's earlier support, we've made a consensus decision to go with: AOM overrides ARIA with no reflection.

This is not set in stone. But we're going to set aside this issue for now, work on implementations, and tighten up the existing spec, hopefully start working on next phases while this one bakes.

**Comment by joanmarie** [comment]
At the end of [Section 2.2 AccessibleNode interface](https://wicg.github.io/aom/spec/#accessiblenode), it says:

> When both an AccessibleNode property and its equivalent ARIA attribute are both present, the ARIA attribute overrides the AccessibleNode property.

Should that be updated? (I see @minorninth's "not set in stone" and "set aside this issue for now" comments, but the spec and explainer have conflicting statements.)
